lib/shared/types/src/types/variableKeys.ts

```diff
@@ -0,0 +1,46 @@
+import { VariableTypeAlias, VariableValue } from './config/models'
+
+/**
+ * Used to support strong typing of variable keys in the SDK.
+ * Usage;
+ * ```ts
+ * import '@devcycle/types';
+ * declare module '@devcycle/types' {
+ *   interface CustomVariableDefinitions {
+ *     'flag-one': boolean;
+ *   }
+ * }
+ * ```
+ * Or when using the cli generated types;
+ * ```ts
+ * import '@devcycle/types';
+ * declare module '@devcycle/types' {
+ *   interface CustomVariableDefinitions extends DVCVariableTypes {}
+ * }
+ * ```
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface CustomVariableDefinitions {}
+type DynamicBaseVariableDefinitions =
+    keyof CustomVariableDefinitions extends never
```

Comment:
How does the `never` work here? How would this statement ever be evaluated?

Reply:
interesting: 

```
1. keyof CustomVariableDefinitions extends never: This expression checks if the CustomVariableDefinitions interface has any keys.
- keyof CustomVariableDefinitions produces a union of the keys of the CustomVariableDefinitions interface. If the interface has no keys, this results in the never type.
- extends never checks if this union is never, which would mean that CustomVariableDefinitions has no keys.
```

Reply:
For some more explanation even though I think you probably got the gist of it from the ChatGPT/Copilot explanation (this was my code from the previous PR);

The interface `CustomVariableDefinitions` can be augmented via a process called "Declaration Merging" which has an [entire page dedicated to it in the TS docs](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) (meaning it's not all that straightforward 😆 ).

So because of this `CustomVariableDefinitions` could be either an empty object if the client code hasn't augmented it, or an object with keys if it has been augmented. Meaning the interface kind of acts like a generic - the non declaration merging way that people would usually do in their own client code would be `type DynamicBaseVariableDefinitions<TDefs> = keyof TDefs extends never ? ...`

Using `CustomVariableDefinitions` on it's own would require users to always do the declaration merging or any "variable key" argument would be defined as `key: never`.
Particularly for those non-typescript users who just use types "invisibly" as IDE autocomplete or just the "automatic get out of my way" typescript users we want the key parameters to be `string` typed.

With all that, you could read the type definition in natural language as;

```
if CustomVariableDefinitions has not been augmented by Declaration merging (ie keys are never)
then VariableDefinitions should be `Record<string, string | number | boolean>`
otherwise use the strict typed version
```